### PR TITLE
Weather Card: Beautify

### PR DESCRIPTION
**Pull request by zsarnett**
### Previous
image

### New

image

## Type of change

-  Dependency upgrade
-  Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
-  Breaking change (fix/feature causing existing functionality to break)
-  Code quality improvements to existing code or addition of tests

## Example configuration

```yaml

- entity: weather.dark_sky
  type: weather-forecast
- type: weather-forecast
  entity: weather.dark_sky
  show_forecast: false
```

## Additional information

- Link to documentation pull request: 

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

[docs-repository]: https://github.com/home-assistant/home-assistant.io

### CODE REVIEW

**Comment by zsarnett** [comment]
https://i.gyazo.com/0babd1b0539196883c8a36653b64c064.gif

**Comment by Misiu** [comment]
What about humidity and wind speed? Are those going to be displayed (if available)?

**Comment by zsarnett** [comment]
Those attributes will be available via more-info window. With all of our cards the goal should not to show all details available but the most important so the card is glancable and more information can be found my clicking the card

**Comment by zsarnett** [comment]
Technically, the attributes will be shown in this priority:

Extrema
Precipitation
Humidity
None

File: src/panels/lovelace/cards/hui-weather-forecast-card.ts [hunk 1]
```diff
@@ -128,20 +104,16 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
...
   protected render(): TemplateResult {
     if (!this._config || !this.hass) {
       return html``;
     }
 
-    const stateObj = this.hass.states[this._config.entity];
+    const stateObj = this.hass.states[this._config.entity] as WeatherEntity;
 
-    if (!stateObj) {
+    if (!stateObj || stateObj.state === UNAVAILABLE) {
```
Comment (zsarnett) [inline src/panels/lovelace/cards/hui-weather-forecast-card.ts:116]: What is the new plan for unavailable? This card doesn't have an input so it doesn't really fall under the same as the light card, etc

Comment (bramkragten) [reply src/panels/lovelace/cards/hui-weather-forecast-card.ts:116]: Just print that it is unavailable

**Comment by zsarnett** [comment]
image

File: src/panels/lovelace/cards/hui-weather-forecast-card.ts [hunk 2]
```diff
@@ -153,140 +127,127 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
...
-                  `
-                : ""}
-              <div class="temp">
-                ${stateObj.attributes.temperature}<span
-                  >${this.getUnit("temperature")}</span
-                >
+          <div class="icon-info">
+            ${stateObj.state in weatherIcons || stateObj.state in weatherImages
+              ? html`
+                  <state-badge
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-weather-forecast-card.ts:166]: Let's not use a `state-badge` if we are going to override everything, just use an `img` or `ha-icon`

File: src/panels/lovelace/cards/hui-weather-forecast-card.ts [hunk 3]
```diff
@@ -153,140 +127,127 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
...
-                    `
-                  )}
-                </div>
-              `
-            : ""}
-        </div>
+                      ${item.condition !== undefined && item.condition !== null
+                        ? html`
+                            <div>
+                              <state-badge
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-weather-forecast-card.ts:221]: Same

File: src/panels/lovelace/cards/hui-weather-forecast-card.ts [hunk 4]
```diff
@@ -295,175 +256,222 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
...
+      return;
     }
-    return ``;
+    this._resizeObserver.observe(card);
   }
 
-  private _showValue(item: string): boolean {
-    return typeof item !== "undefined" && item !== null;
+  private _measureCard() {
+    const card = this.shadowRoot!.querySelector("ha-card");
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-weather-forecast-card.ts:289]: Do we need to measure `ha-card` or we can just measure `this`?

File: src/panels/lovelace/cards/hui-weather-forecast-card.ts [hunk 5]
```diff
@@ -295,175 +256,222 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
...
+      return;
+    }
+    this._narrow = card.offsetWidth < 375;
+    if (card.offsetWidth < 300) {
+      this.setAttribute("verynarrow", "");
+    } else {
+      this.removeAttribute("verynarrow");
+    }
+    if (card.offsetWidth < 200) {
+      this.setAttribute("veryverynarrow", "");
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-weather-forecast-card.ts:300]: Do we need `veryverynarrow`? Can't we just use `verynarrow`?

Comment (zsarnett) [reply src/panels/lovelace/cards/hui-weather-forecast-card.ts:300]: Very very narrow is to deal with times when its in a horizontal stack on a phone. Makes the icon and temp stacked nicely in the center

**Comment by zsarnett** [comment]
Thank you to cgtobi (I cant find your github username) for the new images that don't include a shadow

**Comment by zsarnett** [comment]
@bramkragten This is ready for review.

File: src/panels/lovelace/editor/config-elements/hui-weather-forecast-card-editor.ts [hunk 6]
```diff
@@ -49,10 +49,6 @@ export class HuiWeatherForecastCardEditor extends LitElement
     return this._config!.theme || "";
   }
 
-  get _show_forecast(): boolean {
-    return this._config!.show_forecast || true;
-  }
```
Comment (bramkragten) [inline src/panels/lovelace/editor/config-elements/hui-weather-forecast-card-editor.ts]: This might be still needed for the check in `_valueChanged`?

**Comment by bramkragten** [comment]
Comment on the layout, I think the left and right side should be aligned, now the right is higher than the left.

**Comment by zsarnett** [comment]
Are you talking about height?

**Comment by zsarnett** [comment]
image

File: src/panels/lovelace/cards/hui-weather-forecast-card.ts [hunk 7]
```diff
@@ -92,6 +66,16 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
...
+
+  private _resizeObserver?: ResizeObserver;
+
+  public connectedCallback(): void {
+    super.connectedCallback();
+    this.updateComplete.then(() => this._measureCard());
+  }
+
   public getCardSize(): number {
     return 4;
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-weather-forecast-card.ts:80]: This should change depending on if we show forecast

File: src/panels/lovelace/cards/hui-weather-forecast-card.ts [hunk 8]
```diff
@@ -150,140 +131,139 @@ class HuiWeatherForecastCard extends LitElement implements LovelaceCard {
...
 
+    let hourly: boolean | undefined;
+
+    if (forecast?.length && forecast?.length > 1) {
+      const date1 = new Date(forecast[0].datetime);
+      const date2 = new Date(forecast[1].datetime);
+      const timeDiff = date2.getTime() - date1.getTime();
+
+      hourly = timeDiff < DAY_IN_MILLISECONDS;
+    }
```
Comment (bramkragten) [inline src/panels/lovelace/cards/hui-weather-forecast-card.ts:158]: We should only do this if we going to show the forecast

**Comment by bramkragten** [comment]
In edit mode the card is always rendered as veryverynarrow because the card doesn't get `display: block` from `hui-card-options`

**Comment by bramkragten** [comment]
Is that a weird shadow in the png? 
image

**Comment by cgtobi** [comment]
Great work, my github handle is @cgtobi ;)

**Pull request merged**